File: examples/i2c-mux/coremodel-i2c-mux.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <errno.h>
#include <string.h>


#include "coremodel-i2c-mux.h"
#include <coremodel.h>


#define I2C_MUX_DEBUG 0
#define CHANNEL_LIMIT 8

typedef struct i2c_mux_state{

    void *cm;
    void *handle;
    char busname[32];

    struct i2c_channel{
        struct i2c_coremodel_device{
            char name[32];
            uint8_t addr;
            void *dev_state;
            const coremodel_i2c_func_t *i2c_func;
            void *handle;
            void *mux_state;
            void *channel;
        }device[128];
        uint32_t count;
    }channel[CHANNEL_LIMIT];

    uint8_t count, addr, chan, chanp, write, stop;

} i2c_mux_state_t;

int i2c_mux_connect(void *priv, unsigned channel, unsigned addr, const coremodel_i2c_func_t *dev_func, void *dev_priv, const char *name)
{
    i2c_mux_state_t *state = priv;

    printf("i2c_mux_connect %s addr %x channel %d \n", name, addr, channel);

    if((addr >= 128) || (channel >= CHANNEL_LIMIT) || (state->channel[channel].count >= 128)){
        return 1;
    }

    sprintf(state->channel[channel].device[state->channel[channel].count].name, "%s", name);
    state->channel[channel].device[state->channel[channel].count].addr = addr;
    state->channel[channel].device[state->channel[channel].count].dev_state = dev_priv;
    state->channel[channel].device[state->channel[channel].count].i2c_func = dev_func;
    state->channel[channel].device[state->channel[channel].count].mux_state = state;
    state->channel[channel].device[state->channel[channel].count].channel = &state->channel[channel];
    state->channel[channel].count++;

    return 0;
}

static void i2c_mux_update_channels(i2c_mux_state_t *state)
{
    uint8_t changed = state->chanp ^ state->chan;

    for(int i = 0; i < CHANNEL_LIMIT; i++){
        if(changed & (1 << i)){
            if(state->chan & (1 << i)){
                for(int j = 0; j < state->channel[i].count; j++){

                    state->channel[i].device[j].handle = coremodel_attach_i2c(state->cm, state->busname, state->channel[i].device[j].addr, state->channel[i].device[j].i2c_func,
                                                                              state->channel[i].device[j].dev_state, COREMODEL_I2C_WRITE_ACK);
                    if(!state->channel[i].device[j].handle){
                        printf("failed to attach %s with address %d on channel %d \n", state->channel[i].device[j].name, state->channel[i].device[j].addr, i);
                    }else{
                        printf("attach success\n");
                    }
                }
            }else{
                for(int j = 0; j < state->channel[i].count; j++){

                    if(state->channel[i].device[j].handle){
                        coremodel_detach(state->channel[i].device[j].handle);
                        state->channel[i].device[j].handle = NULL;
                    }else{
                        printf("no handle to detach %s with address %d on channel %d \n", state->channel[i].device[j].name, state->channel[i].device[j].addr, i);
                    }
                }
            }

        }
    }

}

static int i2c_mux_start(void *priv)
{

    printf("START\n");

    return 1;
}

static int i2c_mux_write(void *priv, unsigned len, uint8_t *data)
{
    i2c_mux_state_t *state = priv;
    unsigned i;

    for(i=0; i<len; i++){

#if I2C_MUX_DEBUG
            printf("i2c_mux write data %04x length %d \n", data[i], len);
#endif


        if(state->count == 0){
            state->chanp = state->chan;
            state->write = 1;
        }
        state->chan = data[i];
        state->count++;
    }
    return i;
}

static int i2c_mux_read(void *priv, unsigned len, uint8_t *data)
{
    i2c_mux_state_t *state = priv;
    unsigned i;
    printf("READ [%d]\n", len);

    for(i=0; i<len; i++){
        data[i] = state->chan;
    }

    return i;
}

static void i2c_mux_stop(void *priv)
{
    i2c_mux_state_t *state = priv;

    printf("STOP write %d\n", state->write);

    state->stop = 1;
    state->count = 0;

}

static const coremodel_i2c_func_t i2c_mux_func = {
    .start = i2c_mux_start,
    .write = i2c_mux_write,
    .read = i2c_mux_read,
    .stop = i2c_mux_stop };

const coremodel_i2c_func_t test_i2c_func = {
    .start = test_i2c_start,
    .write = test_i2c_write,
    .read = test_i2c_read,
    .stop = test_i2c_stop };

int main(int argc, char *argv[])
{
    i2c_mux_state_t *state = calloc(1, sizeof(i2c_mux_state_t));
    int res;
    char name[16];

    if(!state){
        return 0;
    }

    if(argc != 3) {
        printf("usage: coremodel-mux <address[:port]> <i2c>\n");
        free(state);
        return 1;
    }

    res = coremodel_connect(&state->cm, argv[1]);
    if(res) {
        fprintf(stderr, "error: failed to connect: %s.\n", strerror(-res));
        free(state);
        return 1;
    }

    sprintf(state->busname, "%s", argv[2]);

    state->handle = coremodel_attach_i2c(state->cm, argv[2], 0x70, &i2c_mux_func, state, 0); //COREMODEL_I2C_WRITE_ACK);
    if(!state->handle) {
        fprintf(stderr, "error: failed to attach i2c mux.\n");
        coremodel_disconnect(state->cm);
        free(state);
        return 1;
    }

    for(int i = 0; i < CHANNEL_LIMIT; i++){
        sprintf(name, "i2c-dummy:%d", i);
        res = i2c_mux_connect(state, i, 0x60 + i, &test_i2c_func, NULL, name);

    }

    while(1){
        coremodel_mainloop(state->cm, 1000000);
        if(state->write && state->stop){
            state->write = 0;
            state->stop = 0;
            i2c_mux_update_channels(state);
        }
    }

    coremodel_detach(state->handle);
    coremodel_disconnect(state->cm);


    free(state);

    return 0;
}

File: examples/i2c-mux/coremodel-i2c-mux.py
# CoreModel I2C Mux Example
#
# Copyright (c) 2026 Corellium Inc.
# SPDX-License-Identifier: Apache-2.0

from coremodel import *

I2C_MUX_DEBUG = 0
CHANNEL_LIMIT = 8

class I2CChannel:
    def __init__(self):
        self.device = [CoreModelI2C] * 128 # [CoreModelI2C] * 128
        self.count = 0
    def __del__(self):
        pass

class I2CMux(CoreModelI2C):
    def __init__(self, busname, address, devname):
        super().__init__(busname = busname, address = address)
        self.devname = devname
        self.flags = COREMODEL_I2C_WRITE_ACK
        self.channel = [I2CChannel] * 8
        for i in range(CHANNEL_LIMIT):
            self.channel[i].__init__(self.channel[i])
            self.channel[i].count = 0
        self.count = 0
        self.addr = 0
        self.chan = 0
        self.chanp = 0
        self.writef = 0
        self.stopf = 0
        self.handle = None
        self.cm = None

    def i2c_mux_connect(self, channel, obj):

        print("i2c_mux_connect {} addr {} channel {} ".format( name, obj.address, channel))

        if (channel >= CHANNEL_LIMIT) or (self.channel[channel].count >= 128):
            return 1
        # print("channel {} channel length {} device len {}".format( channel, len(self.channel), len(self.channel[channel].device)))
        self.channel[channel].device[self.channel[channel].count] = obj
        self.channel[channel].count += 1
        return 0

    def i2c_mux_update_channels(self):

        changed = self.chanp ^ self.chan

        for i in range(CHANNEL_LIMIT):
            if changed & (1 << i) :
                if self.chan & (1 << i):
                    for j in range(self.channel[i].count):

                        self.cm.attach(self.channel[i].device[j])
                        if not self.channel[i].device[j].handle:
                            print("failed to attach {}} with address {} on channel {} ".format( self.channel[i].device[j].name, self.channel[i].device[j].address, i))
                        else:
                            print("attach success")

                else:
                    for j in range(self.channel[i].count):

                        if self.channel[i].device[j].handle:
                            self.cm.detach(self.channel[i].device[j])
                            self.channel[i].device[j].handle = None
                        else:
                            print("no handle to detach {} with address {} on channel {} ".format(self.channel[i].device[j].devname, self.channel[i].device[j].addr, i))


    def start(self):
        print("START")
        return 1

    def write(self, data):
        length_of_data_consumed = len(data)

        for i in range(length_of_data_consumed):

            if I2C_MUX_DEBUG:
                print("i2c_mux write data {} length {}".format( data[i], len))

                if self.count == 0:
                    self.chanp = self.chan
                    self.writef = 1

                self.chan = data[i]
                self.count += 1

            return length_of_data_consumed

        return length_of_data_consumed

    def read(self, data):
        length_of_data_consumed = len(data)
        print("READ {}", len)

        for i in range(length_of_data_consumed):
            data[i] = self.chan
        return length_of_data_consumed

    def __del__(self):
        pass

if __name__ == "__main__":
    if len(sys.argv) != 4:
        print("usage: coremodel-mux <address> <ports> <i2c>")
        sys.exit(1)

    address =  sys.argv[1]
    port = sys.argv[2]
    res = 0
    name = ""

    cm0 = CoreModel("cm0", address, port, "./../../")

    i2cmux0 = I2CMux(sys.argv[3],0x70, "i2c-mux:0")

    cm0.attach(i2cmux0)

    i2c_dummies = [CoreModelI2C] * CHANNEL_LIMIT

    for i in range(CHANNEL_LIMIT):
        name =  "i2c-dummy:{}".format(i)
        print("index {}".format(i))
        i2c_dummies[i] = CoreModelI2C(i2cmux0.name, 0x60 + i)
        i2c_dummies[i].devname = name
        i2cmux0.i2c_mux_connect(i, i2c_dummies[i])

    while(1):
        cm0.mainloop()
        if i2cmux0.writef and i2cmux0.stopf:
            i2cmux0.writef = 0
            i2cmux0.stopf = 0
            i2cmux0.i2c_mux_update_channels()



